### PR TITLE
HIP Compilation Fixes

### PR DESCRIPTION
**Pull request by fthaler**
- Enables HIP-Clang-compilation and verification on full GT, including examples etc.
- Enables GitHub action testing the HIP-Clang builds.
- Removes random default GPU architecture (sm_35), now the architecture has to be specified explicitly (or the compiler default will used).
- Proper host-device overloads for atomic functions.
- Changed example code to be HIP-compatible without compiler-detecting macros.
- Testing environment for Ault using HIP on Vega20.

### CODE REVIEW

**Comment by fthaler** [comment]
jenkins

**Comment by fthaler** [comment]
perftest

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
perftest

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
Waiting for a new compilation of HIP-Clang where a bug with global constexpr variables is fixed…

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -37,6 +37,6 @@ jobs:
       run: |
         python3 pyutils/driver.py -vv build -b ${{ matrix.build_type }} -o $(pwd)/build -i $(pwd)/install -t perftests
     - name: run tests
-      if: (!contains(matrix.compiler, 'cuda')) # no GPUs available -> not tests
+      if: (!contains(matrix.compiler, 'cuda') && !contains(matrix.compiler, 'hip')) # no GPUs available -> not tests
```
Comment (havogt) [inline .github/workflows/tests.yml:40]:
```suggestion
      if: (!contains(matrix.compiler, 'cuda') && !contains(matrix.compiler, 'hip')) # no GPUs available -> no tests
```

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
Test run on Ault on Jenkins finished successfully, but was quite slow (58min). After the merge we can easily enable daily builds (+ daily perftests).

File: include/gridtools/common/cuda_runtime.hpp [hunk 2]
```diff
@@ -11,6 +11,10 @@
 
 #if defined(__HIP__)
 #include <hip/hip_runtime.h>
+#ifdef NDEBUG
```
Comment (anstaf) [inline include/gridtools/common/cuda_runtime.hpp:14]: To be honest, I don't understand why this is not noop. What happens without defining `assert` on our own?

Comment (fthaler) [reply include/gridtools/common/cuda_runtime.hpp:14]: It’s just a workaround for a HIP bug: HIP defines assert independently of `NDEBUG`, leading to compilation errors in our code (as it tries to access some stuff in release mode that is only defined when `NDEBUG` is present). But I actually forgot to file a HIP bug, so thanks for the reminder!

**Comment by fthaler** [comment]
launch jenkins

**Comment by fthaler** [comment]
perftests

**Comment by fthaler** [comment]
Will run perftests, to make sure that CUDA behaves as before (shouldn‘t have changed, but who knows)…

**Pull request merged**